include/validator/validator_component.h

```diff
@@ -17,10 +33,34 @@ struct ExternDescVisitor {
 };
 
 struct InstanceExprVisitor {
-  Expect<void> operator()(const Instantiate &) {
-    // TODO: Validation of instantiate requires each <importname> in c to
-    // match a name in a with argument (compared as strings) and for the
-    // types to match.
+  InstanceExprVisitor(Context &Ctx) : Ctx(Ctx) {}
+
+  Expect<void> operator()(const Instantiate &Inst) {
+    auto Args = Inst.getArgs();
+    auto ImportMap = getImports(Inst.getComponentIdx());
+
+    for (const auto &Arg : Args) {
+      const auto &ArgName = Arg.getName();
+      auto it = ImportMap.find(std::string(ArgName));
+
+      if (it == ImportMap.end()) {
```

Comment:
Here you check if `import xxx ...` has no argument `with xxx ...` is correct. However, at the end of this algorithm, you forget to check import map is empty (no missing argument).

I suggest you reverse the relation: each name of an import statement, find a proper argument.

Reply:
Here, the big picture of component is more clear: It's a module/class, an instantiation is make/new an instance.

If you familiar with OCaml/Agda, module will be easier; if you familiar with Java/C++, class will be easier to understand.

Each import statement, e.g.
```
(component
  (import "T" (type $T (sub resource)))
  ...)
```
means find a component instance named `"T"` in environment, check it's type is a resource. `$T` is omittable, it binds type of `"T"` to use in the component.